obs-webrtc: Add AV1 Support

Sean-Der opened the pull request.

eric committed:
libobs/util: Implement error reader for ffmpeg posix pipe

Transition from using popen() to using fork()/dup2()/execl() to provide
a way to read stderr from ffmpeg.

Co-Authored by: PatTheMav <PatTheMav@users.noreply.github.com>

openglfreak committed:
libobs: Add encode_texture2 function to struct obs_encoder_info

And use it if non-NULL instead of encode_texture.

GeorgesStavracas committed:
linux-pipewire: Add PipeWire-based camera source

This plugin adds a camera source for capture devices available
via PipeWire using the Camera portal.

GeorgesStavracas committed:
linux-pipewire: Allow passing registry events on creation

This will allow consumers of obs_pipewire to handle the device
registry themselves. This is not useful for the screencast code
since there's always only one node in the registry, but it will
be used by the camera and audio code to list hardware devices.

GeorgesStavracas committed:
linux-pipewire: Add obs_pipewire_roundtrip()

Generalize the async server roundtrip wait so that future consumers
can, for example, wait for all devices to be listed before acting
on the registry.

pkviet committed:
obs-ffmpeg: Support multiple audio tracks (mpegts)

This adds support for multiple audio tracks for the new mpegts output.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
UI: Enable audio multi-track w/ mpegts streaming

This enables audio multi-track support in UI for mpegts streams (srt,
rist ...).
The UI changes were coded though to allow re-use by other protocols.

Signed-off-by: pkv <pkv@obsproject.com>

wtay committed:
linux-pipewire: Add NV12 support

List it in addition to YUV formats.

Co-authored-by: Georges Basile Stavracas Neto
<georges.stavracas@gmail.com>

wtay committed:
linux-pipewire: Start camera only after we find the device

Wait with starting the camera until we find the device in the
registry. Avoids cases where the camera stays unlinked.

wtay committed:
linux-pipewire: Collect device controls

Collect the controls on the node. Enumerate the controls as
properties.

Co-authored-by: Georges Basile Stavracas Neto
<georges.stavracas@gmail.com>

wtay committed:
linux-pipewire: Handle property changes

wtay committed:
linux-pipewire: Add resolution and framerate list contents

Trigger a renegotiation of the camera stream when either the
resolution or the framerate change.

wtay committed:
linux-pipewire: Add float property support

wtay committed:
linux-pipewire: Use spa_strerror to convert SPA errors to string

g_strerror only works for errno values, SPA errors are negative errno
values.

wtay committed:
linux-pipewire: Handle camera hotplug

Rebuild the camera list when a camera was added or removed.

kkartaltepe committed:
libobs,libobs-opengl: enable GPU encoding for OpenGL

Enable all of the previously Windows only paths for OpenGL backends that
support encode_texture2

Co-authored-by: Torge Matthies <openglfreak@googlemail.com>

kkartaltepe committed:
libobs-opengl: Enable imported dmabufs for rendering

For now just tag all imported images with GS_RENDER, this may not work
for some images that were produced by some hardware other than the GPU
render engines. But since we don't import VA-API decoded frames we
probably won't run into this. And we need this to render into VA-API
frames destined for encoding.

tuduweb committed:
UI: Add "audio_names" setting to FFmpeg output settings

tuduweb committed:
obs-ffmpeg: Set audio stream title for FFmpeg record output

nowrep committed:
obs-ffmpeg: Add AV1 support for VA-API

jgh-twitch committed:
librtmp: Add missing extended timestamp in Type 3 chunks

According to https://rtmp.veriskope.com/docs/spec/#5313-extended-timestamp
extended timestamps need to be present in Type 3 chunks
(`RTMP_PACKET_SIZE_MINIMUM`) if the previous chunk also included an
extended timestamp

PatTheMav committed:
mac-avcapture: Update plugin to ObjC and modern APIs

Marks old AVFoundation capture source as deprecated.

r-value committed:
libcaption: Optimize branch conditons

The optimization silences the warning about type limits on platforms
with `char` type as `unsigned char`.

The original condition is semantically identical to the optimized one
because the signed-to-unsigned cast is well-defined in C standard.

voluntas commented:
Hi, Thanks for all your great work. I was able to successfully broadcast with my WebRTC SFU 💯 

[![Image from Gyazo](https://i.gyazo.com/923f83c51e145d3bb329be290edda3a9.jpg)](https://gyazo.com/923f83c51e145d3bb329be290edda3a9)

palana committed:
obs-ffmpeg: Add `disable_scenecut` option for NVENC

This is functionally the same as x264's `scenecut=0` option

GeorgesStavracas committed:
linux-pipewire: Rename a function

Make it a bit more semantic, and matching of PipeWire API.

GeorgesStavracas committed:
linux-pipewire: Factor out formats list into a file

This will help share these formats with the upcoming Camera portal
code.

nowrep committed:
obs-ffmpeg: Set better VA-API defaults

Use High profile for H264 and auto level with all codecs.
Remove setting default value for unused "rendermode" option.

derrod committed:
CI: Sign and package Windows builds

derrod committed:
UI: Cancel session end request on commitDataRequest() signal

derrod committed:
libobs/util,docs: Deprecate circlebuf

derrod committed:
libobs/util: Add deque

derrod committed:
libobs: Replace circlebuf with deque

derrod committed:
frontend-tools: Replace circlebuf with deque

derrod committed:
deps/media-playback: Replace circlebuf with deque

derrod committed:
deps/obs-scripting: Replace circlebuf with deque

derrod committed:
obs-ffmpeg: Replace circlebuf with deque

derrod committed:
obs-filters: Replace circlebuf with deque

derrod committed:
obs-outputs: Replace circlebuf with deque

derrod committed:
docs: Add deque documentation

derrod committed:
UI: Save invalid transition source data in safe mode

GeorgesStavracas committed:
linux-pipewire: Pass framerate and resolution at construction

This allows the stream negotiation to set the camera values right
from the start.

derrod committed:
UI: Save last used output resolution to scene collection

jech committed:
obs-webrtc: Improve error messages on HTTP failure

nquah committed:
obs-outputs: Switch RTMP color metadata order

Sends color metadata before the video header for RTMP streams. Only
applicable to HDR RTMP streams. Ensures the server sees the correctly
configured color metadata first, in cases where the decoder config in
the video header has incorrect or missing color information. Mitigates a
bug that users with outdated AMD drivers may encounter.

nquah committed:
obs-outputs: Enable HDR for HEVC over RTMP

Removes flag guard to enable HDR streaming for HEVC over enhanced-RTMP.
This functionality is currently only supported by YouTube. See
github.com/veovera/enhanced-rtmp for the enhanced-RTMP spec.

tt2468 committed:
rtmp-services: Add IRLToolkit Frankfurt ingest

PatTheMav committed:
CI: Update build workflow to use Xcode 15.1 on macOS

derrod committed:
libobs: Add encoder ROI functions

derrod committed:
obs-x264: Implement ROI feature

derrod committed:
obs-ffmpeg: Add ROI support for native NVENC

derrod committed:
obs-ffmpeg: Add ROI support to AMF

derrod committed:
obs-qsv11: Add ROI support

derrod committed:
obs-ffmpeg: Remove Ubuntu 20.04 NVENC fallback

Effectively reverts dd26fe4f8a5fc603a405841bfb02069bf59ee5c8

derrod committed:
obs-ffmpeg: Remove Windows 7 NVENC fallback

Penwy committed:
vlc-video: Fix play_pause to respect boolean

twelho committed:
build-aux: Work around oneVPL search path for Flatpak QSV

QuickSync requires oneVPL, which is unable to find its libraries in the
Flatpak, resulting in MFX_ERR_NOT_FOUND and the encoders not working.
The libraries are present in the install prefix (/app/lib), which is
currently not automatically added to the search paths, see the issue
https://github.com/intel/libvpl/issues/119 for details. For now,
work around the issue by setting ONEVPL_SEARCH_PATH explicitly.

Commit:
mac-virtualcam: Reduce excessive polling for new sample buffers

derrod committed:
libobs: Send item_remove signal after detaching item

derrod committed:
libobs: Fix crop to bounding box not being duplicated with item

Lain-B committed:
libobs: Add a couple missing comments

(Lain note: This is mostly so more (yes more) can be added to this
god-forsaken structure without it getting too messy. In terms of actual
code, I need to be better about writing actual code comments. ...Meaning
that I actually need to start writing code comments.)

Lain-B committed:
libobs: Do not allow controllable media without flag

The OBS_SOURCE_CONTROLLABLE_MEDIA flag is required for controllable
media callbacks to be usable. These checks are being added so we can
defer most controllable media callbacks to the video thread.

Lain-B committed:
libobs/util: Add da_pop_front()

Doesn't really make sense that it was missing, so just add it.

Lain-B committed:
libobs: Defer media controls to video thread

To prevent sources from having to take too much extra consideration into
threading, defer the media control functions that directly affect
functionality to the video thread. Getters will still have to take
threading into account, but this should make things much more trivial
for media controls thread-wise.

(Lain note: Context: I noticed that things such as the slide show
require mutexes due to their media controls, and I felt that it was
largely unnecessary and that libobs should mitigate the threading issue
itself and keep it all in the video thread like it should be. Again, the
getters are still going to require *some* consideration into threading
but in terms of threading, for the type of stuff we're doing, querying
state is far more trivial to take into consideration.)

PatTheMav committed:
mac-capture: Update capture to include menu bar and child windows

macOS 14.2 changed internal defaults of SCK requiring the inclusion
of the menu bar and child windows to be enabled explicitly.

This will have a slight negative impact on capture performance but is
required to restore behavior of prior macOS versions.

PatTheMav committed:
cmake: Update build requirements to Xcode 15.1 and macOS 14.2 SDK

Lowest supported macOS version to build(!) macOS will be 13.5 with this
change.

derrod committed:
UI: Switch to bigger units for high bitrate/large recordings

RytoEX committed:
CI: Update build workflow comment to Xcode 15.1

Amends 389cb95e15dd708be2d4b46e8190d60dfe4d07ea.

rhutsAMD committed:
obs-ffmpeg: Fix issue with B-frames introducing motion blur

If a user sets both AdaptiveMiniGOP=true and EnablePreAnalysis=true
in the AMF/FFmpeg options field, AMF will adaptively insert
B-pictures, and no longer uses the fixed B pattern.

For a fixed B-frames pattern, it is expected that increasing B-frames
can cause a quality drop for certain content such as with high motion.
AdaptiveMiniGOP is recommended when using B-frames to improve the
quality in such cases. AdaptiveMiniGOP is dependent on PreAnalysis
which means that trying to enable it without having PreAnalysis turned
ON will have no negative effect (AdaptiveMiniGOP won't be enabled).

tytan652 committed:
UI: Fix unused-variable

Some variables were not guarded for being unused when built with FFmpeg 6.1

PatTheMav committed:
obs-ffmpeg: Disable hardened runtime when using ad-hoc signing

sebastian-s-beckmann committed:
UI: Use normal source width for screenshots

Fixes an issue where filters that resize a source, like "Crop/Pad" or
"Scaling/Aspect Ratio", change the texture but the new source size gets
ignored, effectively either cropping the screenshot or giving it black
borders.

Commit:
libobs: Log Mac model identifier

Commit:
libobs: Use aligned memory allocator in Cocoa logging

Sean-Der committed:
obs-webrtc: Add AV1 Support

PatTheMav committed:
cmake: Update MbedTLS finder to support MSVC Debug configuration

Natively compiled MbedTLS for Windows will use the non-debug platform
library, which will lead to a linker error when building OBS in debug
configuration with MSVC.

Neutralizing the linker directive for Debug builds on Windows
unbreaks the build configuration without the need to ship separate
debug variants of MbedTLS.

RytoEX committed:
obs-ffmpeg: Fix incorrect CMake list syntax

sebastian-s-beckmann committed:
mac-capture: Remove macOS 14 ifdef and custom CGDisplayStream header

ec9809bd43e3571fd60d6950eb54a4cc2ad13aa2 increased the minimum macOS SDK
to 14.2, meaning that the ifdef (and as a result the custom header) is
no longer needed.

derrod committed:
obs-ffmpeg: Only use NVENC compatibility hack if necessary

derrod committed:
libobs: Pass parent pointer into signal_item_remove

With aaf8d8b38fba518d6dbeef92fbf75b0738843d10 changing the order
item->parent would be NULL thus resulting in a crash.

derrod committed:
obs-ffmpeg: Fix NVENC compat hack ternary being inverted

Lain-B committed:
libobs/util: Fix function declaration

C functions that have no parameters should use (void) to signify that
they have no parameters, otherwise legacy C handling for parameters will
apply.

Lain-B committed:
image-source: Rework slideshow (slideshow mk2)

Makes it so that slideshows can have as many files as desired, and
lazily decodes image data in a separate thread to prevent decoding from
stalling the video thread.

rhutsAMD committed:
obs-ffmpeg: Default the number of B-frames to 3

jpark37 committed:
libobs-d3d11: Remove new/delete replacements

MSVC noticed inline is not allowed, and new/delete replacments also
don't belong in a header file. We're also just leaning on
__STDCPP_DEFAULT_NEW_ALIGNMENT__ being 16 for MSVC x64 now.

derrod committed:
UI: Add and migrate to GPU scaling options

derrod committed:
docs: Add encoder ROI functions/struct

norihiro committed:
cmake: Fix Xcode version error message

tytan652 committed:
cmake: Skip generate export header if interface in legacy path

It backports a behavior from CMake 3.0 to CMake 2.0

tytan652 committed:
cmake: Skip PDB install for interface target

Lain-B committed:
Merge pull request #10039 from derrod/roi-libobs

libobs/plugins: Add region of interest (ROI) encoder feature

derrod committed:
image-source: Migrate slideshow mk2 to deque

tytan652 committed:
build-aux: Update Flatpak oneVPL module to libvpl

Update oneVPL v2023.3.1 to libvpl v10.1

tytan652 committed:
build-aux: Update KDE Flatpak runtime to 6.6

This new runtime version relies on Freedesktop SDK 23.08.

Disable PipeWire module since Freedesktop SDK 23.08 provides a more
recent version of PipeWire.

Intel Media SDK is patch to support being compiled with more recent
version of GCC

tytan652 committed:
build-aux: Update Flatpak modules

* Update SVT-AV1 to 1.8.0
* Update aom to 3.8.0
* Update libsrt to 1.5.3
* Update nv-codec-headers to 12.1.14.0
* Update FFmpeg to 6.1.1
* Update LuaJIT to 2.1 c525bcb902
* Update Asio to 1.29.0
* Update nlohmann JSON to 3.11.3
* Update Intel oneVPL GPU to 23.4.3
* Update libdatachannel to 0.19.5
  * Update plog to 1.1.10

RytoEX committed:
obs-ffmpeg: Use 2 tile columns in NVENC AV1 for 4K+

When resolution is 4K or higher, use 2 tile columns for NVENC AV1.

RytoEX committed:
obs-qsv11: Use 2 tile columns in QSV AV1 for 4K+

When resolution is 4K or higher, use 2 tile columns for QSV AV1.

RytoEX committed:
CI: Update deps to 2024-01-11 release

Notable changes:
 * deps.qt: Fix library naming for macOS debug builds
 * deps.ffmpeg: Fix local universal architecture builds of aom for macOS
 * deps: Update Windows build script with explicit package names
 * deps.ffmpeg: Switch Windows builds to native build toolchain

Lain-B committed:
Merge pull request #9943 from derrod/its-a-deque

Rename `circlebuf` to `deque`

RytoEX committed:
libobs: Use std _Pragma with MSVC

Instead of using the Microsoft-specific __pragma keyword, use the
standard _Pragma directive, which is now supported in C11 and C++11
/std modes.

RytoEX committed:
CI: Update deps to 2024-01-18 release

To avoid a compiler error, this commit also bumps the configured NVENC
version to match the updated version in the deps package.

Notable changes:
 * deps.ffmpeg: Update zlib to 1.3
 * deps.ffmpeg: Update SVT-AV1 to 1.8.0
 * deps.ffmpeg: Update aom to 3.8.0
 * deps.ffmpeg: Update libsrt to 1.5.3
 * deps.ffmpeg: Update nv-codec-headers to 12.1.14.0
 * deps.ffmpeg: Update AMF to 1.4.32
 * deps.ffmpeg: Update FFmpeg to 6.1
 * deps.macos: Update LuaJIT to 2.1 c525bcb902
 * deps.macos: Update FreeType to 2.13.2
 * deps.macos: Update Asio to 1.29.0
 * deps.macos: Update nlohmann JSON to 3.11.3
 * deps.macos: Update Sparkle to 2.5.2
 * deps.macos: Update Syphon Framework to 5.0 39e31383ff
 * deps.windows: Update FreeType to 2.13.2
 * deps.windows: Update curl to 8.5.0
 * deps.windows: Update LuaJIT to 2.1 c525bcb902
 * deps.windows: Update Asio to 1.29.0
 * deps.windows: Update nlohmann JSON to 3.11.3
 * deps.windows: Update VPL to v2.10.1
 * deps.qt: Update Qt6 to 6.6.1 for Windows
 * deps.qt: Update Qt6 to 6.6.1 for macOS

Be1erafon committed:
rtmp-services: Remove StreamVi (#10130)

derrod committed:
libobs: Fix audio doubling with multiple main view mixes

RytoEX committed:
obs-browser: Update version to 2.23.3

7fa3822 - REALLY don't allow browser docks to arbitrarily close OBS

exeldro committed:
libobs: Fix GPU scaling using the main mix

hhgyu committed:
Add PandaTV live RTMP service (#10117)

RytoEX committed:
obs-websocket: Update version to 5.4.0

See b61a5c243137681a206bca2cef0f92e20598353a for changelog.

DriftedClouds committed:
UI: Add accessible names to hotkey edit boxes

RytoEX committed:
CI: Update deps to 2024-01-27 release

This commit also modifies build-aux to update libdatachannel.

This commit also modifies obs-webrtc to avoid a compiler error.

Notable changes:
 * deps.ffmpeg: Update libdatachannel to 0.20.1
 * deps.macos: Remove Carla
 * deps.macos: Add uthash to macOS deps
 * deps.windows: Remove Carla
 * deps.windows: Add uthash to Windows deps

kkartaltepe committed:
deps/glad: Add EGL_EXT_device_base related extensions

kkartaltepe committed:
libobs,libobs-opengl,libobs-d3d11: Add opengl gs_enum_adapters

This adds gs_enum_adapters and gs_get_adapter_count to the opengl
backend and promotes these to multiplatform graphics functions.

However we need to make an internal device change, device_enum_adapters
must pass in the current device on opengl to ensure that adapter #0 is
the display adapter. We do this to avoid changes to plugins already
checking against obs_video_info.adapter which is always 0 and expected
to be the device OBS was initialized on.

The actual implementation reports the dri render node (or /Software).
This allows plugins to query non-video features of the adapters like
VA-API/NVENC/etc or other cross device functionality. `/Software` is
chosen to avoid opening random files in the current directory if its
passed along as a file path like the regular dri render nodes.

tytan652 committed:
obs-webrtc: Require libdatachannel 0.20 or newer

kkartaltepe committed:
obs-qsv11: Use the new gs_enum_adapters

Sean-Der commented:
AV1 has been merged into master
